### ISSUE TITLE
[packagers] Flatpak OOB when using BINARY tar.gz archive

### ISSUE THREAD

**Issue by crschnick** [jreleaser/jreleaser#1324]
[trace.log](https://github.com/jreleaser/jreleaser/files/11237999/trace.log)

The actual artifacts (If needed to reproduce): https://github.com/xpipe-io/xpipe_staging/releases/tag/0.5.36

**Comment by aalmiray** [comment]
The expectation for  Zip and Tar distributions is that the root entry matches the archive filename, such that`xpipe-portable-linux-x86_64.tar.gz` should have a root entry named `xpipe-portable-linux-x86_64`. However, in this case the root entry is `xpipe-0.5.36`.

See https://jreleaser.org/guide/latest/concepts/distributions/binary.html#_creating_a_distribution

**Comment by crschnick** [comment]
I see. How strict is this requirement in terms of the possibility to remove this? In my opinion this kinda negates the point of the `transform` attribute of distributions. Because you have to know the final name when you package it and are therefore not allowed to transform it.

**Comment by aalmiray** [comment]
Indeed. The `archive` assembler may be used to create the archive as its name accepts name templates. But if the archive is created by external means then the problem persists. It may be possible to inspect an archive whose root entry has a different value than the expected one but I'm sure other packagers will continue to fail. For example, Docker and Macports  still expect the root entry to match the archive name.

The code needed to solve this issue is not trivial because of the different places that need to be updated but it's possible.

**Comment by crschnick** [comment]
Oh well that sounds complicated. I can just create another jreleaser distribution specifically for packagers as I don't want to change my naming scheme for my other releases.

I guess jreleaser should warn/error when such an invalid archive name is detected.

**Comment by aalmiray** [comment]
Perhaps. However, I think relaxing the expectations would also be a good thing to do. I'll run a code spike and see how far I can get.

**Comment by aalmiray** [comment]
One more thing, it's also expected that artifacts define a version in their names. `xpipe-portable-linux-x86_64.tar.gz` does not define a version. This could also cause trouble down the line. Would `xpipe-portable-0.5.36-linux-x86_64.tar.gz` be acceptable?

**Comment by aalmiray** [comment]
Found one more thing. The unixArchive inspector expects a single distribution with a `bin` directory. xpipe configures 2 applications as part of the distribution: `app` and `cli`. This results in binary files being copied without executable permissions:

```
 - install -Dm644 app/bin/xpiped /app/app/bin/xpiped
 - install -Dm644 cli/bin/xpipe /app/cli/bin/xpipe
```

**Comment by aalmiray** [comment]
I have a fix for arbitrary binaries in flatpak and spec. Still, compound distributions won't work with other packagers. Flatpak descriptor for xpipe should find the following binaries

```
 - install -Dm755 app/bin/xpiped /app/app/bin/xpiped
 - install -Dm755 app/lib/runtime/bin/java /app/app/lib/runtime/bin/java
 - install -Dm755 app/lib/runtime/bin/keytool /app/app/lib/runtime/bin/keytool
 - install -Dm755 app/lib/runtime/bin/rmiregistry /app/app/lib/runtime/bin/rmiregistry
 - install -Dm755 app/lib/runtime/bin/xpiped /app/app/lib/runtime/bin/xpiped
 - install -Dm755 cli/bin/xpipe /app/cli/bin/xpipe
```

**Comment by crschnick** [comment]
> One more thing, it's also expected that artifacts define a version in their names. `xpipe-portable-linux-x86_64.tar.gz` does not define a version. This could also cause trouble down the line. Would `xpipe-portable-0.5.36-linux-x86_64.tar.gz` be acceptable?

Where exactly is that requirement coming from? The reason I don't include the version is that it makes it easier to handle the artifacts in other scenarios like in a workflow file where I can just refer to them via the fixed name.

> Found one more thing. The unixArchive inspector expects a single distribution with a `bin` directory. xpipe configures 2 applications as part of the same distribution: `app` and `cli`. This results in binary files being copied without executable permissions:
> 
> ```
>  - install -Dm644 app/bin/xpiped /app/app/bin/xpiped
>  - install -Dm644 cli/bin/xpipe /app/cli/bin/xpipe
> ```
> 
> Technically these 2 applications should be distributed with separate archives. There's no current workaround for packagers to support this layout. Have to think for a moment how to expose such configuration as some packagers may be OK with this kind of "compound" distribution (flatpak, spec) but others may not (sdkman).

The main reason for this separation was that some users confused the executables when using the portable distribution. I guess for packagers, I can merge these two directories into one such that there is only one bin directory.

**Comment by aalmiray** [comment]
> Where exactly is that requirement coming from? The reason I don't include the version is that it makes it easier to handle the artifacts in other scenarios like in a workflow file where I can just refer to them via the fixed name. 

The expectation from JReleaser based on common practices for naming artifacts is `<artifactId>-<version>[-classifier]`. In this case `artifactId` is just a name (thus 'xpipe-portable' is fine) while `classifier` matches the target platform.

> The main reason for this separation was that some users confused the executables when using the portable distribution. I guess for packagers, I can merge these two directories into one such that there is only one bin directory.

Based on what I can gather by looking at the files, `app` is a desktop application that bundles a Java runtime (built with jpackage) while `cli` is a command line application compiled to native code using GraalVM native (?). If `app` and `cli` have no dependency to one another that is, you may install x-pipe/app separate from x-pipe/cli, then they technically should have separate distributions (archives) for NATIVE_INSTALLER (app) and BINARY (cli). In the case of `flatpak`, the app manifest is different for desktop apps than for cli apps (see `stereotype` property).

Now, I understand is convenient for users to download a single archive that delivers both apps, both default templates for package managers as currently set by JReleaser expect a single distribution type per archive. It may be possible to support multiple distributions but you'll have to provide your own templates for the time being. If this were really to be needed then the `template generate` command will help you get started.

**Comment by aalmiray** [comment]
For more reference, https://github.com/jreleaser/jreleaser/tree/main/core/jreleaser-templates/src/main/resources/META-INF/jreleaser/templates/binary/flatpak defines the set of possible templates used by the flatpak packager for both `cli` and `desktop` stereotypes of a BINARY distribution.

The key difference is the `.desktop` file and an extra `<launchable>` element in the metainfo file.

Moreover, it's possible to bundle multiple binaries in the same distribution and let flatpak know about them (see https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-provides) but the default JReleaser template does not handle that case (it expects a single binary at the moment). This could be added to JReleaser as well. Filed https://github.com/jreleaser/jreleaser/issues/1326

**Comment by crschnick** [comment]
> > Where exactly is that requirement coming from? The reason I don't include the version is that it makes it easier to handle the artifacts in other scenarios like in a workflow file where I can just refer to them via the fixed name.
> 
> The expectation from JReleaser based on common practices for naming artifacts is `<artifactId>-<version>[-classifier]`. In this case `artifactId` is just a name (thus 'xpipe-portable' is fine) while `classifier` matches the target platform.

I can look into adapting that naming scheme for the artifacts, that shouldn't be an impossible task.

> > The main reason for this separation was that some users confused the executables when using the portable distribution. I guess for packagers, I can merge these two directories into one such that there is only one bin directory.
> 
> Based on what I can gather by looking at the files, `app` is a desktop application that bundles a Java runtime (built with jpackage) while `cli` is a command line application compiled to native code using GraalVM native (?). If `app` and `cli` have no dependency to one another that is, you may install x-pipe/app separate from x-pipe/cli, then they technically should have separate distributions (archives) for NATIVE_INSTALLER (app) and BINARY (cli). In the case of `flatpak`, the app manifest is different for desktop apps than for cli apps (see `stereotype` property).

The app and cli actually have a dependency on each other. The app can't be compiled using GraalVM native image as it dynamically loads modules at runtime. The cli is therefore separate to decrease start up time, and handle the instance management of the daemon. You can interact with the daemon through the cli, e.g. the command to start the app would be `xpipe open`, that is also the launch command that I use in the `.desktop` file.

I guess my project does not fit into the standard distribution definitions, so I will see how I can work around that.

**Comment by aalmiray** [comment]
> The app and cli actually have a dependency on each other. The app can't be compiled using GraalVM native image as it dynamically loads modules at runtime. The cli is therefore separate to decrease start up time, and handle the instance management of the daemon. You can interact with the daemon through the cli, e.g. the command to start the app would be xpipe open, that is also the launch command that I use in the .desktop file.
>
> I guess my project does not fit into the standard distribution definitions, so I will see how I can work around that.

Gotcha. In this case a single archive as you have it today is the way to go. Which means the default template provided by flatpak may fall short of what you need. I'd recommend generating a local copy of the templates and adapt them to your needs. Make sure to set `desktop` as stereotype for the distribution (you could also set it at the project level).

`gradle jreleaserTemplateGenerate --distribution-name portable --packager-name flatpak`.

**Comment by crschnick** [comment]
So to avoid further problems, should I move the portable distributions to the jreleaser files type instead if they don't fit the usual criteria for any distribution type? And then create a separate packager distribution that somehow fits the criteria.

**Comment by aalmiray** [comment]
It depends. If you’d like to configure additional packagers such as homebrew, sdkman, scoop, and perhaps chocolatey then yes, a single distribution type per archive would be recommended _if_ using the default templates provided by JReleaser. 

But xpipe’s case is different as `app` depends on `cli` for better user experience. If separate distros and archives were to be published now users would have to install 2 distros instead of just one. 

JReleaser will support multiple binaries per distribution in the future. That should help you avoiding custom templates for most cases. 

Now for Homebrew, I think I’d be best if the `app` distro were to be packaged as a `.dmg` or `.pkg` which should make the formula contain a cask (that’s how UI apps are installed in osx) while the `cli` could remain as a regular formula. In which case 2 distinct formulas will be needed, with `xpipe` (app) depending in `xpipe-cli` (cli), such that installing the former also installs the latter while installing the latter does no install the former (assuming cli may be used in standalone mode).

Long story short, I think keeping the portable archive as you have it to be installed via flatpak is OK. But also using a custom template to expose 2 binaries instead of just the one. Do not modify the archive nor split into several distros unless you want to explore other packagers where the aforementioned restrictions apply. Hopefully JReleaser will provide what’s needed in a timely manner.

**Comment by aalmiray** [comment]
🎉 This issue has been resolved in `v1.6.0` ([Release Notes](https://github.com/jreleaser/jreleaser/releases/tag/v1.6.0))
